File: lib/services/game_service/models/sensor_data.dart

/// A class representing sensor data associated with a game node.
///
/// [SensorData] includes various environmental readings and status indicators
/// that influence the player's decisions and the game's progression.
class SensorData {
  /// The current radiation level detected, represented as a string descriptor.
  ///
  /// Examples include "Low", "Moderate", "High", etc.
  final String? radiationLevel;

  /// Additional sensor readings or environmental features.
  ///
  /// This map can include various other sensor data points as needed.
  final Map<String, dynamic>? additionalData;

  /// Creates a new instance of [SensorData].
  ///
  /// - [batteryLevel]: The current battery level percentage.
  /// - [radiationLevel]: Optional radiation level descriptor.
  /// - [additionalData]: Optional map of additional sensor data.
  SensorData({
    this.radiationLevel,
    this.additionalData,
  });

  /// Creates a [SensorData] instance from a JSON map.
  ///
  /// This factory constructor is useful for deserializing JSON data into
  /// `SensorData` objects.
  ///
  /// - [json]: A `Map<String, dynamic>` containing the sensor data.
  ///
  /// Returns a new instance of [SensorData].
  factory SensorData.fromJson(Map<String, dynamic> json) {
    return SensorData(
      radiationLevel: json['radiation_level'] as String?,
      additionalData: json['additional_data'] != null
          ? Map<String, dynamic>.from(json['additional_data'] as Map)
          : null,
    );
  }

  /// Converts the [SensorData] instance into a JSON map.
  ///
  /// This method is useful for serializing `SensorData` objects into JSON format.
  ///
  /// Returns a `Map<String, dynamic>` representing the sensor data.
  Map<String, dynamic> toJson() {
    return {
      if (radiationLevel != null) 'radiation_level': radiationLevel,
      if (additionalData != null) 'additional_data': additionalData,
    };
  }
}

File: ios/WatchApp Watch App/Models/SensorData.swift
import Foundation

/// A class representing sensor data associated with a game node.
///
/// `SensorData` includes various environmental readings and status indicators
/// that influence the player's decisions and the game's progression.
class SensorData: Equatable {
    /// The current radiation level detected, represented as a string descriptor.
    ///
    /// Examples include "Low", "Moderate", "High", etc.
    let radiationLevel: String?
    
    /// Additional sensor readings or environmental features.
    ///
    /// This dictionary can include various other sensor data points as needed.
    let additionalData: [String: Any]?
    
    /// Creates a new instance of `SensorData`.
    ///
    /// - Parameters:
    ///   - radiationLevel: Optional radiation level descriptor.
    ///   - additionalData: Optional dictionary of additional sensor data.
    init(radiationLevel: String?, additionalData: [String: Any]?) {
        self.radiationLevel = radiationLevel
        self.additionalData = additionalData
    }
    
    /// Creates a `SensorData` instance from a JSON dictionary.
    ///
    /// - Parameter json: A `[String: Any]` dictionary containing the sensor data.
    convenience init?(json: [String: Any]) {
        // Extract `radiationLevel` as a `String?`
        let radiationLevel = json["radiation_level"] as? String
        
        // Extract `additionalData` as `[String: Any]?`
        let additionalData = json["additional_data"] as? [String: Any]
        
        // Initialize using the primary initializer
        self.init(radiationLevel: radiationLevel, additionalData: additionalData)
    }
    
    /// Converts the `SensorData` instance into a JSON dictionary.
    ///
    /// - Returns: A `[String: Any]` dictionary representing the sensor data.
    func toJson() -> [String: Any] {
        var json = [String: Any]()
        
        if let radiationLevel = radiationLevel {
            json["radiation_level"] = radiationLevel
        }
        
        if let additionalData = additionalData {
            json["additional_data"] = additionalData
        }
        
        return json
    }
    
    /// Compares two `SensorData` instances for equality.
    static func == (lhs: SensorData, rhs: SensorData) -> Bool {
        // Compare `radiationLevel` directly
        guard lhs.radiationLevel == rhs.radiationLevel else { return false }
        
        // Compare `additionalData` dictionaries manually
        guard lhs.additionalData?.count == rhs.additionalData?.count else { return false }
        for (key, lhsValue) in lhs.additionalData ?? [:] {
            // Ensure the key exists in `rhs.additionalData`
            guard let rhsValue = rhs.additionalData?[key] else { return false }
            
            // Convert both values to `String` for comparison (if possible)
            if String(describing: lhsValue) != String(describing: rhsValue) {
                return false
            }
        }
        
        return true
    }
}
